File: Java/app/src/main/java/org/geeksforgeeks/demo/AudioRecordingAdapter.java
package org.geeksforgeeks.demo;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

// Adapter to display and manage a list of audio recordings
public class AudioRecordingAdapter extends RecyclerView.Adapter<AudioRecordingAdapter.AudioViewHolder> {

    private final List<AudioRecording> recordings;
    private final OnPlayClickListener onPlayClick;
    private final OnDeleteClickListener onDeleteClick;
    private final OnShareClickListener onShareClick;

    public AudioRecordingAdapter(List<AudioRecording> recordings,
                                 OnPlayClickListener onPlayClick,
                                 OnDeleteClickListener onDeleteClick,
                                 OnShareClickListener onShareClick) {
        this.recordings = recordings;
        this.onPlayClick = onPlayClick;
        this.onDeleteClick = onDeleteClick;
        this.onShareClick = onShareClick;
    }

    // ViewHolder class holds the views for each list item
    public static class AudioViewHolder extends RecyclerView.ViewHolder {
        ImageView playButton;
        TextView fileNameText;
        SeekBar seekBar;
        ImageView deleteButton;
        ImageView shareButton;

        public AudioViewHolder(@NonNull View itemView) {
            super(itemView);
            playButton = itemView.findViewById(R.id.playPauseButton);
            fileNameText = itemView.findViewById(R.id.fileName);
            seekBar = itemView.findViewById(R.id.seekBar);
            deleteButton = itemView.findViewById(R.id.delete);
            shareButton = itemView.findViewById(R.id.share);
        }
    }

    // Inflates the layout for each item in the list
    @NonNull
    @Override
    public AudioViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_audio_recording, parent, false);
        return new AudioViewHolder(view);
    }

    // Binds data to each item and sets up click listeners
    @Override
    public void onBindViewHolder(@NonNull AudioViewHolder holder, int position) {
        AudioRecording recording = recordings.get(position);

        holder.fileNameText.setText(recording.getFileName());

        // Update play button based on playback state
        holder.playButton.setImageResource(
                recording.isPlaying() ? R.drawable.pause_square : R.drawable.play_square
        );

        // Play button click triggers callback
        holder.playButton.setOnClickListener(v -> onPlayClick.onPlayClick(recording, holder.getAdapterPosition()));

        // Delete button click triggers callback
        holder.deleteButton.setOnClickListener(v -> onDeleteClick.onDeleteClick(recording, holder.getAdapterPosition()));

        // Share button click triggers callback
        holder.shareButton.setOnClickListener(v -> onShareClick.onShareClick(recording));

        // Reset seekBar to the beginning for now
        holder.seekBar.setProgress(0);
    }

    // Returns the size of the dataset
    @Override
    public int getItemCount() {
        return recordings.size();
    }

    // Interfaces for click callbacks
    public interface OnPlayClickListener {
        void onPlayClick(AudioRecording recording, int position);
    }

    public interface OnDeleteClickListener {
        void onDeleteClick(AudioRecording recording, int position);
    }

    public interface OnShareClickListener {
        void onShareClick(AudioRecording recording);
    }
}

File: Kotlin/app/src/main/java/org/geeksforgeeks/demo/AudioRecordingAdapter.kt
package org.geeksforgeeks.demo

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

// Adapter to display and manage a list of audio recordings
class AudioRecordingAdapter(
    private val recordings: MutableList<AudioRecording>,
    private val onPlayClick: (AudioRecording, Int) -> Unit,
    private val onDeleteClick: (AudioRecording, Int) -> Unit,
    private val onShareClick: (AudioRecording) -> Unit
) : RecyclerView.Adapter<AudioRecordingAdapter.AudioViewHolder>() {

    // ViewHolder class holds the views for each list item
    inner class AudioViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val playButton: ImageView = itemView.findViewById(R.id.playPauseButton)
        val fileNameText: TextView = itemView.findViewById(R.id.fileName)
        val seekBar: SeekBar = itemView.findViewById(R.id.seekBar)
        val deleteButton: ImageView = itemView.findViewById(R.id.delete)
        val shareButton: ImageView = itemView.findViewById(R.id.share)
    }

    // Inflates the layout for each item in the list
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AudioViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_audio_recording, parent, false)
        return AudioViewHolder(view)
    }

    // Binds data to each item and sets up click listeners
    override fun onBindViewHolder(holder: AudioViewHolder, position: Int) {
        val recording = recordings[position]

        holder.fileNameText.text = recording.fileName

        // Update play button based on playback state
        holder.playButton.setImageResource(
            if (recording.isPlaying) R.drawable.pause_square else R.drawable.play_square
        )

        // Play button click triggers callback
        holder.playButton.setOnClickListener {
            onPlayClick(recording, holder.adapterPosition)
        }

        // Delete button click triggers callback
        holder.deleteButton.setOnClickListener {
            onDeleteClick(recording, holder.adapterPosition)
        }

        // Share button click triggers callback
        holder.shareButton.setOnClickListener {
            onShareClick(recording)
        }

        // Reset seekBar to the beginning for now
        holder.seekBar.progress = 0
    }

    // Returns the size of the dataset
    override fun getItemCount(): Int = recordings.size
}